### ISSUE TITLE
:AIChat hangs and has no output for higher number of tokens

### ISSUE THREAD

**Issue by Konfekt** [madox2/vim-ai#140]
I did not pinpoint the exact number of tokens passed, but for example for a visual selection of 800 lines with < 6k words, `:'<,'>AIChat /role` hangs for a few minutes and then shows no output on Vim 9.1.919 on Ubuntu 22.04.
Using e.g. an OpenAI model this should still be below theirs maximal number of tokens (around 200k for some models).

**Comment by madox2** [comment]
Is this still an issue?

**Comment by Konfekt** [comment]
Yes, it's simply very slow. For example, I fed lines containing each its number from 1 to 1000 and it takes a minute to respond. Whereas if you paste them into the aichat buffer, the response is immediate. Likely the copying of the range is slow. It's fast with 100, takes a bit with 500 and very long with 1000 lines.

**Comment by madox2** [comment]
I visually selected 1000 lines (Tokens 8,607, Characters 44750) and it went all good, got response from models (gpt-4o, gemini) almost immediately.

Could you try to debug where is the plugin stuck most of the time in your case?

Could you also share how many tokens are causing this issue? (https://platform.openai.com/tokenizer)

**Comment by Konfekt** [comment]
Found it:

```vim
" ftplugin/aichat.vim (lines 130-137)
augroup Aichat
  autocmd! * <buffer>
  autocmd BufWinEnter <buffer> call s:MarkdownRefreshSyntax(1)
  autocmd BufUnload <buffer> call s:MarkdownClearSyntaxVariables()
  autocmd BufWritePost <buffer> call s:MarkdownRefreshSyntax(0)
  autocmd InsertEnter,InsertLeave <buffer> call s:MarkdownRefreshSyntax(0)
  autocmd CursorHold,CursorHoldI <buffer> call s:MarkdownRefreshSyntax(0)
augroup END
```

changes the syntax on every InsertEnter/Leave and there was a syntax autocmd.

**Comment by madox2** [comment]
Still cannot reproduce, could be something specific about your environment? Maybe it could collide with some other plugins

**Comment by Konfekt** [comment]
Inside ftplugin/aichat.vim

```vim
autocmd vimrcFileTypeAIChat Syntax <buffer> setlocal foldexpr=AIChatFold() foldmethod=expr
function! AIChatFold() abort
  return getline(v:lnum) =~# '^>>> user$' ? '>1' : '='
endfunction
```

In theory, `BufWinenter ++once` should suffice, but in practice the syntax scripts of the embedded file types keep overriding the fold settings, see https://github.com/vim/vim/issues/16083 , so Syntax is the safer variant.

However,  why does the Syntax have to be refreshed on switching between normal and insert mode?

**Comment by madox2** [comment]
There is no folding in vim-ai, are you using some patch version or extending plugin in some way?

**Comment by Konfekt** [comment]
Of course, this is a setting to reproduce.

**Comment by Konfekt** [comment]
Could these autocmd be made optional as they can slow down and should not be necessary ?

```vim
augroup Aichat
  autocmd! * <buffer>
  autocmd BufWinEnter <buffer> call s:MarkdownRefreshSyntax(1)
  autocmd BufUnload <buffer> call s:MarkdownClearSyntaxVariables()
  autocmd BufWritePost <buffer> call s:MarkdownRefreshSyntax(0)
  autocmd InsertEnter,InsertLeave <buffer> call s:MarkdownRefreshSyntax(0)
  autocmd CursorHold,CursorHoldI <buffer> call s:MarkdownRefreshSyntax(0)
augroup END
```

**Comment by madox2** [comment]
> Could these autocmds be made optional as they can slow down Vim and should not be necessary for syntax highlighting ?

I am not sure if this is true. I used your script for folding and found the bottleneck in how initial promt was being inserted. I optimized it and it looks to be working fine. For now I keep it on the branch: https://github.com/madox2/vim-ai/tree/big-prompt-rendering-optimization please have a look

**Comment by Konfekt** [comment]
Oh, I missed that. Now it's merged.And it's lighting fast now. Thank you!

**Comment by Konfekt** [comment]
For reference,

```vim
def s:AIChatFold(): string
  return getline(v:lnum) ==# '>>> user' ? '>1' : '='
enddef

autocmd vimrcFileTypeAIChat Syntax <buffer> setlocal foldexpr=s:AIChatFold() foldmethod=expr
doautocmd vimrcFileTypeAIChat Syntax <buffer>
```

speeds up folding
